src/charm.py

```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charmed Machine Operator for the PostgreSQL database."""
+
+import logging
+import secrets
+import string
+import subprocess
+from typing import List
+
+from charms.operator_libs_linux.v0 import apt
+from ops.charm import ActionEvent, CharmBase
+from ops.main import main
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    Relation,
+    WaitingStatus,
+)
+
+from cluster import (
+    ClusterAlreadyRunningError,
+    ClusterCreateError,
+    ClusterNotRunningError,
+    ClusterStartError,
+    PostgresqlCluster,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class PostgresqlOperatorCharm(CharmBase):
+    """Charmed Operator for the PostgreSQL database."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self._postgresql_service = "postgresql"
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(
+            self.on.get_postgres_password_action, self._on_get_postgres_password
+        )
+        self._cluster = PostgresqlCluster(self._get_postgresql_version())
```

Comment:
You essentially have a factory pattern here, but I think your factory is in the wrong place. Rather than include the `_get_postgresql_version` in the charm itself, it might make more sense to have a getPostgresqlCluster() method which introspects aspects of the system and returns the correct PostgresqlCluster object.

The only real reason you have the get_postgresql_version is for version strings in file paths, which could probably be contained within the PostgresqlCluster class itself; removing the need for a factory altogether.

Reply:
I agree, Billy! Thanks for catching that. I also use the version in the `apt.add_package` call (which didn't worked with the package `postgresql`, without the version - I'll check the call to make sure it's correct), which can also be retrieved with your suggestion about `apt-cache` in the comment below. I'll change that to get the version in PostgresqlCluster object, which makes more sense.

Reply:
I removed the factory. I also updated the apt library in order to install the postgresql package instead of the ones with the version in the name.